Add FAQ section in docs (#768)
minio/directpv: docs/faq.md

FAQ
----

#### What type of disks are recommended for DirectPV?

> DirectPV is specifically meant for [Direct Attached Storage](https://en.wikipedia.org/wiki/Direct-attached_storage) such as hard drives, solid-state drives and JBODs etc. Whereas, SAN and NAS based storage options are not recommended as it involves extra network hops in the data path. This may lead to poor perfomance and increased complexity.

Reviewer: Shall we be more forceful here?

```suggestion
DirectPV is specifically meant for [Direct Attached Storage](https://en.wikipedia.org/wiki/Direct-attached_storage) such as hard drives, solid-state drives and JBODs etc. 

Avoid using DirectPV with SAN and NAS based storage options, as they inherently involve extra network hops in the data path. This leads to poor performance and increased complexity.
```

#### How DirectPV is different from LocalPV and HostPath?

> Hostpath volumes are ephemeral and are tied to the lifecyle of pods. ie, the volumes are lost when a pod is restarted or deleted. Whereas, DirectPV volumes are persistent and can sustain node and pod reboots. And the lifecycle of a DirectPV volume is managed by [Persistent Volume Claims](https://kubernetes.io/docs/concepts/storage/persistent-volumes/).

Reviewer: ```suggestion
Hostpath volumes are ephemeral and are tied to the lifecyle of pods. Hostpath volumes are lost when a pod is restarted or deleted, resulting in the loss of any stored data. 

DirectPV volumes are persistent through node and pod reboots. The lifecycle of a DirectPV volume is managed by the associated [Persistent Volume Claims](https://kubernetes.io/docs/concepts/storage/persistent-volumes/).
```

> LocalPV is statically provisioned where the volumes need to be pre-provisioned before the workloads using them are created. DirectPV is similar to localPV but it supports dynamic provisioning.

Reviewer: ```suggestion
LocalPVs are statically provisioned from a local storage resource on the nodes where persistent volumes are required, and must be created prior to the workloads using them. User management is required down to creating the actual object resource for the local PV.

DirectPV also creates statically provisioned volume resources, but does not require user intervention for the creation of the object resource. Instead, DirectPV dynamically provisions the persistent volume in response to a PVC requesting a DirectPV storage volume. This significantly reduces complexity of management.

```

#### How are the disks selected for a Pod?

> DirectPV selects the disks local to the node where the pod is scheduled. This provides direct access for the pods to the disks. DirectPV runs a selection algorithm to choose a disk for a volume creation request and then creates a sub-directory for the volume with quota set on the sub-directory for the requested volume size and publishes the volume to the pod. To know more on the selection algorithm, please refer [here](./volume-scheduling.md).

Reviewer: ```suggestion
DirectPV operates only on those disks which it [explicitly manages](./cli.md#initialize-the-available-drives-present-in-the-cluster)

1. DirectPV selects managed disks local to the node where the pod is scheduled. This provides direct access for the pods to the disks. 

2. DirectPV runs a selection algorithm to choose a disk for a volume creation request

 3. DirectPV then creates a sub-directory for the volume with quota set on the sub-directory for the requested volume size
 
 4. DirectPV publishes the volume to the pod. 
 
 To know more on the selection algorithm, please refer [here](./volume-scheduling.md).
```

#### What does drive initialization do?

DirectPV command `kubectl directpv init` command will prepare the drives by formatting them with XFS filesystem and mounting them in a desired path (/var/lib/directpv/mnt/<uuid>). Upon success, these drives will be ready for volumes to be scheduled and you can see the initialized drives in `kubectl directpv list drives`.

#### What are the conditions that a drive must satisfy to be used for DirectPV?

A drive must meet the following requirements for it to be used for initialization

- The size of the drive should not be less then 512MiB.
- The drive should not be hidden. (Check /sys/class/block/<device>/hidden, It should not be "1")
- The drive should not be a read-only.
- The drive should not be partitioned.
- The drive should not be held by other devices. (Check /sys/class/block/<device>/holders, It should not have any entries)
- The drive should not be mounted.
- The drive should not have swap-on enabled.

> - The drive should not be a CDROM.

Reviewer: Replace "should not" with "must not" if htese are requirements.

Reviewer: Also wondering if we should state the drive must not be an LVM, LUN, or similar virtualized drive.

Author: we do support LVMs,LUKS and LUNs @ravindk89 It still works with DirectPV

Just that we don't recommend them for production setups.

Reviewer 2: Technically we support any block device, but we do not recommend network drives, LVMs, etc. Hence LUKS on top of DAS is right way to use, we support it.

Reviewer: In that case maybe below this block state

"DirectPV can support any storage resource presenting itself as a block device. While this includes LVMs, LUKs, LUNs, and similar virtualized drives, we do not recommend using DirectPV with those in production environments without SUBNET support."

Or something to that effect.

Author: Added this @ravindk89 

"DirectPV can support any storage resource presenting itself as a block device. While this includes LVMs, LUKs, LUNs, and similar virtualized drives, we do not recommend using DirectPV with those in production environments."
